libyul/optimiser/ASTWalker.h

```diff
@@ -102,4 +102,35 @@ class ASTModifier
 	}
 };
 
+namespace detail
+{
+template <
+	typename Node,
+	typename Visitor,
+	typename Base = std::conditional_t<std::is_const_v<Node>, ASTWalker, ASTModifier>
+>
+struct ForEach: Base
+{
+	template<typename Callable>
+	ForEach(Callable&& _visitor): visitor(std::forward<Callable>(_visitor)) {}
+
+	using Base::operator();
+	void operator()(Node& _node) override
+	{
+		visitor(_node);
+		Base::operator()(_node);
+	}
+
+	Visitor visitor;
+};
+}
+
+/// Helper function that traverses the AST and calls the visitor for each
+/// node of a specific type.
+template<typename Node, typename Entry, typename Visitor>
+void forEach(Entry&& _entry, Visitor&& _visitor)
+{
+	detail::ForEach<Node, std::decay_t<Visitor>>{std::forward<Visitor>(_visitor)}(std::forward<Entry>(_entry));
```

Comment:
Actually, looking at this once more, my comment about ``std::decay_t`` was wrong...
If ``Visitor`` resolves to a reference type, then that reference is sure to be alive during the body of ``forEach`` - so it is actually *good* if ``detail::ForEach`` has a reference member and the constructor does not need to copy. I'll quickly check if this is indeed what will happen and will probably create a followup PR to remove it.